### DISCUSSION TITLE
Google SAML does not work after updating to 2.0.33

### DISCUSSION THREAD

**Discussion by laurensk** [kimai/kimai#4311] [General]
### Describe the issue

**Edit: Down below is my initial problem interpretation. It turned out that the SAML problem started after updating to 2.0.33 and has nothing to do with Cloudflare or any other infrastructure changes.**

Currently, our Google SAML integration does not work any more after enforcing https-only traffic in Cloudflare. It seems that Kimai sets the RelayState url of the saml request to the http:// url of our Kimai installation instead of using the https:// scheme. I have already tried setting the `baseurl` property of the saml config, but this does not change this behaviour. Is there a way to change the content of the RelayState url or is this hard-coded inside the Kimai source (or possibly even in php-saml)?

### I already tried

- [x] I've read and searched [the documentation](https://www.kimai.org/documentation/).
- [X] I've searched for similar issues in this repository.
- [X] I've searched for similar issues in [the discussions](https://github.com/kimai/kimai/discussions).

### Kimai version

2.0.33

### How do you run Kimai?

Virtual Server or alike

### Which PHP version are you using?

8.1

### Logfile

no log entries

**Comment by kevinpapst** [comment]
> enforcing https-only traffic in Cloudflare

What does that mean? Does PHP see that you are using https? If you terminate SSL before and the webserver still sees port 80, it will always use http.

**Reply by laurensk** [reply]
We terminate SSL on our reverse proxy, which is running inside another VM. I get that PHP sees its traffic as HTTP, but when I set the base url to use the https:// schema manually, this value should be used instead of the scheme of the request.

I guess as a workaround I could encrypt the connection between the reverse proxy and Kimai using a self-signed certificate, but as we use the same approach (SAML with HTTP reverse proxy and manually setting the base url) with many other tools successfully, I don't think this is expected behaviour.

I would suggest using the scheme of the baseurl set in the saml config if it's set. If not, just fall back to the current "auto-detecting" behaviour.

~LK

**Reply by laurensk** [reply]
Ok I have created a self signed cert for Kimai, and the RelayState is now correctly set to the HTTPS scheme. But SAML still does not work. This is now a different problem (https://github.com/kimai/kimai/discussions/4312)

**Reply by kevinpapst** [reply]
Hm, I updated many Kimai system from 1.x to 2.x using Azure and/or Google SAML. They kept on working.

As you changed much more on your infrastructure as well, the culprit is likely somewhere else, what we don't see yet.

Did you check, that you really have all your proxy headers in place?
https://www.kimai.org/documentation/webserver-configuration.html#reverse-proxy

**Reply by laurensk** [reply]
Yes, I have set up everything correctly. I also did not change anything on the reverse proxy nor Kimai.
I have included 2 screenshots of how the login process (after the redirect from Google to Kimai) looks like.

<img width="2056" alt="Screenshot 2023-09-21 at 10 38 49" src="https://github.com/kimai/kimai/assets/39766854/78ebcfcf-7b18-4f3c-889b-b740c7a20ec6">

<img width="2047" alt="Screenshot 2023-09-21 at 10 38 54" src="https://github.com/kimai/kimai/assets/39766854/8d256b37-2a8f-43ef-a791-1bfcac618007">

**Reply by kevinpapst** [reply]
Make the setup simpler.
Use a barebone system with nothing in between, test SAML. 
Works? Great, add more complexity. See where/when it fails. Now you know where to look at.

**Reply by laurensk** [reply]
This is currently not possible using our infrastructure. I could install a separate VM and test this, but before I do, I just wanted to ask whether is behaviour (in the screenshots) is normal.

**Reply by kevinpapst** [reply]
See: SAML works for many users, all Cloud customers and in all customer environments where I configured it in the past.
But, as you know, server environments can be complicated and there are many small things that can go wrong.

>  I also did not change anything on the reverse proxy nor Kimai.
> our Google SAML integration does not work any more after enforcing https-only traffic in Cloudflare

I am happy to jump on a paid 1:1 if you want that and we debug the hell out of it.
But I don't think that I can help you remotely with this issue.

**Reply by laurensk** [reply]
I have removed the new Cloudflare infrastructure entirely now - but still - the error persists. I think it really is a problem since the update to 2.0.33

**Reply by laurensk** [reply]
I have restored a backup of 2.0.30 and it works again. So this must be due to a change between 2.0.30 and 2.0.33!

**Reply by laurensk** [reply]
Edit: Improved choice of words

Just to help anyone debugging this issue in case it happens again, in the old Kimai (v2.0.30) the Location header of the acs request is set to the baseurl but in v2.0.33 (as seen in the screenshots I posted previously) the Location header is set to `/en/login`.

So I think the acs request does not correctly set the session as authenticated.

<img width="2056" alt="Screenshot 2023-09-21 at 20 06 29" src="https://github.com/kimai/kimai/assets/39766854/53c785e8-5331-4ef7-9a68-e2794d65eb73">

**Reply by kevinpapst** [reply]
This is the entire changelog your are talking about:
https://github.com/kimai/kimai/compare/2.0.30...2.0.33

This is the only related change I can see that has something to-do with SAML:
https://github.com/kimai/kimai/compare/2.0.30...2.0.33#diff-8f4d7ec1b98e0ef063a86a96af29b4665d6823db1745a9ed97ed8a76109da749

I appreciate every user of Kimai. But I am not here to debug issues from single users.
As long as it works for many other users, nobody else complains and I cannot reproduce it, you are the only one who can debug that.

**Reply by laurensk** [reply]
Should have written "help debugging" instead of "help you debug".
Was not meant to ask you for debugging it, I just posted the screenshot so that when others have an issue with it, there is a baseline on where to start looking or identifying the bug/problem/incident.

**Reply by kevinpapst** [reply]
Hey, there is actually a bug... sorry for not finding earlier.
Read on here: https://github.com/kimai/kimai/discussions/4300
